presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java

```diff
@@ -0,0 +1,285 @@
+
+    private final long wallStart;
+    private final long cpuStart;
+    private final long userStart;
+
+    private long intervalWallStart;
+
+    private long intervalUnaccountedCpuWallTime;
+    private long intervalCpuStart;
+    private long intervalUserStart;
+
+    private OperationTiming overallTiming;
+    private final Map<OperationTiming, InternalTiming> timings = new HashMap<>();
+
+    public OperationTimer()
+    {
+        this(Ticker.systemTicker(), DEFAULT_MIN_NANOS_FOR_CPU_TIMING);
+    }
+
+    public OperationTimer(Ticker ticker, long minNanosForCpuTiming)
+    {
+        this.ticker = requireNonNull(ticker, "ticker is null");
+        this.minNanosForCpuTiming = minNanosForCpuTiming;
+
+        wallStart = ticker.read();
+        cpuStart = currentThreadCpuTime();
+        userStart = currentThreadUserTime();
+
+        intervalWallStart = wallStart;
+        intervalCpuStart = cpuStart;
+        intervalUserStart = userStart;
+    }
+
+    public void setOverallTimer(OperationTiming overallTiming)
+    {
+        requireNonNull(overallTiming, "overallTiming is null");
+        checkState(this.overallTiming == null, "Overall timer already set");
+        this.overallTiming = overallTiming;
+    }
+
+    public void recordOperationComplete(OperationTiming timing)
+    {
+        requireNonNull(timing, "timing is null");
+        InternalTiming internalTiming = timings.computeIfAbsent(timing, InternalTiming::new);
+
+        long intervalWallEnd = ticker.read();
+        long intervalWallDuration = intervalWallEnd - intervalWallStart;
+        // only measure cpu time if the wall interval is large
+        if (intervalWallDuration < minNanosForCpuTiming) {
+            long intervalCpuEnd = currentThreadCpuTime();
+            long intervalCpuDuration = max(0, intervalCpuEnd - intervalCpuStart - intervalUnaccountedCpuWallTime);
+
+            long intervalUserEnd = currentThreadUserTime();
+            long intervalUserDuration = max(0, intervalUserEnd - intervalUserStart - intervalUnaccountedCpuWallTime);
+
+            internalTiming.recordMeasuredCpuTime(intervalWallDuration, intervalCpuDuration, intervalUserDuration);
+
+            intervalUnaccountedCpuWallTime = 0;
+            intervalWallStart = intervalWallEnd;
+            intervalCpuStart = intervalCpuEnd;
+            intervalUserStart = intervalUserEnd;
+        }
+        else {
+            internalTiming.recordUnaccountedCpuTime(intervalWallDuration);
+
+            intervalUnaccountedCpuWallTime += intervalWallDuration;
+            intervalWallStart = intervalWallEnd;
+        }
+    }
+
+    public void end()
+    {
+        if (overallTiming == null && timings.isEmpty()) {
+            return;
+        }
+
+        long wallEnd = ticker.read();
+
+        long cpuEnd;
+        long userEnd;
+        // remeasure cpu timings if the current interval is large
+        if (wallEnd - intervalWallStart < minNanosForCpuTiming) {
```

Comment:
I think we should keep overall cpu time to be always precise. It shouldn't hurt, as that's what we have right now.

Reply:
If you decided to keep the estimation - i think you should flip the condition.